### DISCUSSION TITLE
Section 3 : Unable to find the .pandoc file on my route directory,I make the file by myself but didn't work.

### DISCUSSION THREAD

**Discussion by RaferGuo** [Cyfrin/security-and-auditing-full-course-s23#207] [Q&A]
near this time in video https://youtu.be/pUWmJ86X_do?t=15075

### **Problem**
when I use 
```
pandoc report.md -o report.pdf --from markdown --template=eisvogel --listings
```
to create the pdf, i get “_cound‘t find data file_”.
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/99655c43-74c9-4159-ab72-e5621b915854)

I'm sure i have install pandoc and latex.
use ```sudo apt-get install pandoc```
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/17862efd-178d-44a6-9d34-750b71039f09)
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/81710125-5ded-4088-aea9-729a2e652b02)

I can’t find the doc .pandoc as patrick shows on my route directory, 
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/227bdf21-a668-4a3a-ad89-9e5d73bf75c3)
so i have to mkdir by myself.
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/9f12a666-458d-4f04-b515-ff73ce1245c4)

### **What I have tried.**
so I go back, and use mkdir to create every file the pic shows" /usr/share/pandoc/data/templates/eisvogel.latex".
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/62fbebb0-37f7-4242-aa7a-943859d0bf37)
and i copy the content on https://github.com/Cyfrin/audit-report-templating/blob/main/eisvogel.latex
to make eisvogel.latex
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/450bb700-f072-47ac-a2ae-c0bf2db46405)

it also shows that pandoc direction maybe is this way,
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/b7062206-d343-4cf6-8816-844fa3f0bea4)
so i also make new pandoc document under share document which doesn't exist before. and finaly get like this
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/5e610a4f-9439-4b4d-9fe0-d6965bcf3598)

i use this 2 ways one is get in other Q&A to make pdf，and it still dosen’t work.
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/3c22e613-c43b-49c9-a460-931136ee2efa)

I spent plenty of hours on this problem， so what can i do now，appreciate a lot.

**Comment by EngrPips** [comment]
Hello @RaferGuo, Please refer to the written lesson for that lesson, and you should try to follow the suggested solution in the written lesson`. Let us know if that helps.

**Reply by RaferGuo** [reply]
Actually，at first time， I have already follow the written lesson in this page https://github.com/Cyfrin/audit-report-templating. 
and I didn‘t see the .pandoc folder at the home directory，
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/e3205267-ab6d-4997-b5cf-3d95936d384e)

So I mkdir as below.
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/371da17a-6dec-4955-be80-e65a6aa988cc)
and then it still dosen't work.
![image](https://github.com/Cyfrin/security-and-auditing-full-course-s23/assets/31519376/21a1fa08-02b9-4940-9db5-cebbe4c3b57c)

**Reply by EngrPips** [reply]
> Actually，at first time， I have already follow the written lesson in this page https://github.com/Cyfrin/audit-report-templating. and I didn‘t see the .pandoc folder at the home directory， ![image](https://private-user-images.githubusercontent.com/31519376/346128378-e3205267-ab6d-4997-b5cf-3d95936d384e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxODkxNTgsIm5iZiI6MTcyMDE4ODg1OCwicGF0aCI6Ii8zMTUxOTM3Ni8zNDYxMjgzNzgtZTMyMDUyNjctYWI2ZC00OTk3LWI1Y2YtM2Q5NTkzNmQzODRlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA1VDE0MTQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMzMzQ2YTJlY2E5NzFlMWM0M2RlZDg4OWQyYmViMjBjMGRmNTJhMWFlMWUwOGI2YzM0NjA0ZTIwNzFlMzUzNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HK1iuBLCiVw5Pv4ipQOT1LOqgGSupH0Yayp-mka5z7k)
> 
> So I mkdir as below. ![image](https://private-user-images.githubusercontent.com/31519376/346128881-371da17a-6dec-4955-be80-e65a6aa988cc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxODkxNTgsIm5iZiI6MTcyMDE4ODg1OCwicGF0aCI6Ii8zMTUxOTM3Ni8zNDYxMjg4ODEtMzcxZGExN2EtNmRlYy00OTU1LWJlODAtZTY1YTZhYTk4OGNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA1VDE0MTQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyYzY3MGZiMTRmYTE5NGU4NjVlYzYyOTBhMGE1YTQ0YzAxNjdhMDk5ZmUyYTk1MWQ5MDdhMmU2OTliMmY3ZDImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.vmYCnpg-EHoa2DK3gTGkg3Ve8MVorH3agGy-SctzA4I) and then it still dosen't work. ![image](https://private-user-images.githubusercontent.com/31519376/346130384-21a1fa08-02b9-4940-9db5-cebbe4c3b57c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAxODkxNTgsIm5iZiI6MTcyMDE4ODg1OCwicGF0aCI6Ii8zMTUxOTM3Ni8zNDYxMzAzODQtMjFhMWZhMDgtMDJiOS00OTQwLTlkYjUtY2ViYmU0YzNiNTdjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA1VDE0MTQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1MTQ4YjI0NzNlMjQ3ZDRjZjBkOTY2MzNlMjJmYWFmNWY0NTE0YjVmYmZkNjIyMTMyOGMwZDRhMTU5NTVlODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.kkjrF4-LXliNZy5v3A_V5nwebix9aZSCQOlRiDpo9gU)

Did you try to run the command to make the pdf after following the written lesson instructions before you went ahead to create the directory yourself? Honestly I had issue with this stuff but then i cleared everything deleted everything i had installed and then restarted the process and luckily it worked.

**Comment by 14himanshi** [comment]
Hello , I am also getting same issue @EngrPips 
And I tried to access the written lesson, but it won't load only
can you paste the lesson here?

**Reply by EngrPips** [reply]
---
title: Your First Full Report - Making a PDF
---

_Follow along with this video:_

---

### First Professional Markdown Report

This lesson covers how to convert a list of findings into a professional-looking PDF using **Markdown**.

Our goal is to transform raw data into valuable information by creating a detailed and comprehensive report. Plus, this gives you something impressive to add to your portfolio!

## The Basics

There are some tools and resources you'll need to prepare yourself with before getting started.

[**GitHub Repo**](https://github.com/Cyfrin/audit-report-templating) - We've created a repo dedicated to assisting security reviewers with generating these reports.

[**Pandoc**](https://pandoc.org/installing.html) - a universal document converter that we'll be leveraging to generate our PDFs

[**LaTeX**](https://www.latex-project.org/get/) - a document preparation system for typesetting used in technical and scientific documentation primarily.

[**Markdown All in One**](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) - Amazing VS Code extension to get the most our of markdown formatting.

[**VSCode PDF**](https://marketplace.visualstudio.com/items?itemName=tomoki1207.pdf) - will allow us to preview PDF files within VSCode

### Adding LaTex to Pandoc

Once `Pandoc` has been installed, it should create a folder in your root directory named `.Pandoc`, within is a `templates` folder. We want to navigate there.

In our provided GitHub Repo, you'll find a specific template file named [`eisvogel.latex`](https://github.com/Cyfrin/audit-report-templating/blob/main/eisvogel.latex). You want to copy this file into your `templates` folder.

> This `eisvogel.latex` template is what's going to tell `Pandoc` how to format our PDF for us! Challenge yourself to customize this template in future!

### Setting Up

Once `Pandox` and `LaTex` have been installed, create a file named `report.md` in your audit-data folder.

Within the aforementioned GitHub Repo, you'll find `report-example.md`. Copy this into your newly created file. This will be our template for building our final report.

### Adding Your Own Logo

Lastly, let's add a bit of flare. Find an awesome logo (pdf format) and add it to the audit-data folder as well. Name this file `logo.pdf`.

### Filling out report.md

Inside our `report.md` template, we're going to want to personalize a number of things.

- **Title:** Name it something that describes your work precisely such as "Network Vulnerability Assessment".
- **Author:** You!
- **Date:** Update the audit date.

Now, let's move to the sections under `===` which you can customize according to your audit:

- **Prepared by:** You!
- **Auditors:** You again! If you're working as part of a team, you can list contributors here.
- **Protocol summary:** Describe the protocol and its workings.
- **Disclaimer:** Enter your name in the space provided, this is to assure the protocol knows that the report is not a guarantee of bug-free code.
- **Risk classifications:** Explain the criteria for classifying severities into High, Medium and low.
- **Audit details:** Include the commit hash that your findings correspond to.
- **Scope:** Include reference to the exact contracts the review has covered.
  - _Note:_ the `└── `, found in the README scope will error when we generate the PDF. Replace this with `#--`.
- **Audit roles:** The roles of the protocol, these were some of the earliest notes we took!
- **Executive summary:** Give a brief overview of the assessment process.
- **Severity and number of issues found:** Summarize the number and severity of issues detailed in the report.
- **Findings:** This is our breakdown of specific findings uncovered over the course of the audit. Paste the write-ups we've done into the respective severity categories and delete the ones we don't need!

Our report is now ready to be transformed into a professional looking PDF!

### Generating the PDF

Alright, moment of truth. In your terminal, navigate to your `audit-data` folder. Assuming everything has gone well upto now we should just have to run the command:

```bash
pandoc report.md -o report.pdf --from markdown --template=eisvogel --listings
```

And with a bit of magic, you should see a `report.pdf` file appear in your `audit-data` folder.

### Wrap Up

Wow! Our report looks amazing. It's so professional, any client we provide this to would be impressed. We absolutely should add this to our portfolio to showcase all we've learned. Let's go over that in the next lesson!

---

Ok, this wasn't easy and there are admittedly a tonned of potential pitfalls along the way. I've compiled a few possible errors/scenarios you may run into with some suggestions to troubleshoot them below.

<details>
<summary>Errors/Issues</summary>

1. **My home/root directory doesn't have a `.pandoc` file!**

   - Depending on your operating system, this file may exist elsewhere. If you're using WSL/Linux keep a few things in mind

     - The file may be hidden - files prepended with `.` are often hidden. You can reveal all files in a directory with the command `ls -a`
     - The file may be elsewhere - navigate back in directories (`cd ..`) until you reach one that looks like this

     <img src="/security-section-3/28-making-a-pdf/making-a-pdf1.png" style="width: 75%; height: auto;">

     ...from here navigate to `usr/share/pandoc/data/templates`. In here you will find existing templates and this is where `eisvogel.latex` should be added.

2. **VS Code says I'm _unable to write a file to that directory_!**

   - This is related to your user permissions, we can force the file to be created with a sudo command. `sudo touch eisvogel.latex` - this command will create a file named `eisvogel.latex` in your current directory.
     - You may be prompted to enter your credentials or need to create an admin user.

3. **VS Code says I'm _unable to write to eisvogel.latex_!**

   - Similarly to above, this is permissions related. The easiest work around I found was through another `sudo` command.
     ```bash
     sudo tee eisvogel.latex << 'EOF'
     [copy LaTex here]
     EOF
     ```
   - The LaTex you need to copy is available [**here**](https://github.com/Cyfrin/audit-report-templating/blob/main/eisvogel.latex). Yes, you will be pasting 1068 lines into your terminal - this will overwrite your `eisvogel.latex` file, in your current directory, with that copied data.

4. **When I run `pandoc report.md -o ... etc` I get _File Not Found_**

   - This seems caused when our LaTex package is missing an important element. The easiest solution is to assure we have the full distribution of the package we're using. For WSL users `sudo apt install texlive-full` will resolve these errors.
     - Note: `texlive-full` is 5.6GB in size.

5. **When I run `pandoc report.md -o ... etc` I get _Missing number, treated as zero_**

   - Caused by an error in the LaTex syntax either in your markdown using it, or the template itself. Replace the block of LaTeX at the top of your `report.md` file with the following:

   ```
    \begin{titlepage}
    \centering
    {\Huge\bfseries Protocol Audit Report\par}
    \vspace{2cm}
    \begin{figure}[h]
    \centering
    \includegraphics[width=0.5\textwidth]{logo.pdf}
    \end{figure}
    \vspace{2cm}
    {\Large Version 1.0\par}
    \vspace{1cm}
    {\Large\itshape equious.eth\par}
    \vfill
    {\large \today\par}
    \end{titlepage}
   ```

   This should resolve the error.

**Reply by EngrPips** [reply]
> ## title: Your First Full Report - Making a PDF
> _Follow along with this video:_
> 
> ### First Professional Markdown Report
> This lesson covers how to convert a list of findings into a professional-looking PDF using **Markdown**.
> 
> Our goal is to transform raw data into valuable information by creating a detailed and comprehensive report. Plus, this gives you something impressive to add to your portfolio!
> 
> ## The Basics
> There are some tools and resources you'll need to prepare yourself with before getting started.
> 
> [**GitHub Repo**](https://github.com/Cyfrin/audit-report-templating) - We've created a repo dedicated to assisting security reviewers with generating these reports.
> 
> [**Pandoc**](https://pandoc.org/installing.html) - a universal document converter that we'll be leveraging to generate our PDFs
> 
> [**LaTeX**](https://www.latex-project.org/get/) - a document preparation system for typesetting used in technical and scientific documentation primarily.
> 
> [**Markdown All in One**](https://marketplace.visualstudio.com/items?itemName=yzhang.markdown-all-in-one) - Amazing VS Code extension to get the most our of markdown formatting.
> 
> [**VSCode PDF**](https://marketplace.visualstudio.com/items?itemName=tomoki1207.pdf) - will allow us to preview PDF files within VSCode
> 
> ### Adding LaTex to Pandoc
> Once `Pandoc` has been installed, it should create a folder in your root directory named `.Pandoc`, within is a `templates` folder. We want to navigate there.
> 
> In our provided GitHub Repo, you'll find a specific template file named [`eisvogel.latex`](https://github.com/Cyfrin/audit-report-templating/blob/main/eisvogel.latex). You want to copy this file into your `templates` folder.
> 
> > This `eisvogel.latex` template is what's going to tell `Pandoc` how to format our PDF for us! Challenge yourself to customize this template in future!
> 
> ### Setting Up
> Once `Pandox` and `LaTex` have been installed, create a file named `report.md` in your audit-data folder.
> 
> Within the aforementioned GitHub Repo, you'll find `report-example.md`. Copy this into your newly created file. This will be our template for building our final report.
> 
> ### Adding Your Own Logo
> Lastly, let's add a bit of flare. Find an awesome logo (pdf format) and add it to the audit-data folder as well. Name this file `logo.pdf`.
> 
> ### Filling out report.md
> Inside our `report.md` template, we're going to want to personalize a number of things.
> 
> * **Title:** Name it something that describes your work precisely such as "Network Vulnerability Assessment".
> * **Author:** You!
> * **Date:** Update the audit date.
> 
> Now, let's move to the sections under `===` which you can customize according to your audit:
> 
> * **Prepared by:** You!
> * **Auditors:** You again! If you're working as part of a team, you can list contributors here.
> * **Protocol summary:** Describe the protocol and its workings.
> * **Disclaimer:** Enter your name in the space provided, this is to assure the protocol knows that the report is not a guarantee of bug-free code.
> * **Risk classifications:** Explain the criteria for classifying severities into High, Medium and low.
> * **Audit details:** Include the commit hash that your findings correspond to.
> * **Scope:** Include reference to the exact contracts the review has covered.
>   
>   * _Note:_ the `└── `, found in the README scope will error when we generate the PDF. Replace this with `#--`.
> * **Audit roles:** The roles of the protocol, these were some of the earliest notes we took!
> * **Executive summary:** Give a brief overview of the assessment process.
> * **Severity and number of issues found:** Summarize the number and severity of issues detailed in the report.
> * **Findings:** This is our breakdown of specific findings uncovered over the course of the audit. Paste the write-ups we've done into the respective severity categories and delete the ones we don't need!
> 
> Our report is now ready to be transformed into a professional looking PDF!
> 
> ### Generating the PDF
> Alright, moment of truth. In your terminal, navigate to your `audit-data` folder. Assuming everything has gone well upto now we should just have to run the command:
> 
> ```shell
> pandoc report.md -o report.pdf --from markdown --template=eisvogel --listings
> ```
> 
> And with a bit of magic, you should see a `report.pdf` file appear in your `audit-data` folder.
> 
> ### Wrap Up
> Wow! Our report looks amazing. It's so professional, any client we provide this to would be impressed. We absolutely should add this to our portfolio to showcase all we've learned. Let's go over that in the next lesson!
> 
> Ok, this wasn't easy and there are admittedly a tonned of potential pitfalls along the way. I've compiled a few possible errors/scenarios you may run into with some suggestions to troubleshoot them below.
> 
> Errors/Issues

@14himanshi  Is this what you are asking for?

**Reply by 14himanshi** [reply]
yes , thank you :) @EngrPips

**Reply by EngrPips** [reply]
Anytime, Buddy.

**Comment by afrazahmad-hub** [comment]
Hello everyone,
Someone can help me to get rid of following error.
This error is coming from my `report.md` file.
Thanks

```
Error producing PDF.
! LaTeX Error: No counter 'none' defined.

See the LaTeX manual or LaTeX Companion for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.573 \begin{longtable}[]{@{}lllll@{}}
```

**Reply by EngrPips** [reply]
Have you checked the written lesson to see if this issue is listed and whether there is a suggested solution?

**Reply by afrazahmad-hub** [reply]
Yes, I had multiple issues, and solved following the written lesson and Error/Issues section mentioned above.
But this one is confusing i can't handle it.

**Reply by EngrPips** [reply]
The below is the response I got from Gemini, Please try the suggested solutions and let's see if you are able to resolve the issue 

This error is a known issue that appeared in **Pandoc versions 3.8.1 and 3.8.2**. It typically happens when you have a table in your Markdown file and are converting it to PDF using a LaTeX engine.

### Why this is happening

In recent updates, Pandoc changed how it handles table numbering in LaTeX to prevent tables from being double-counted or numbered incorrectly when using filters like `pandoc-crossref`.

To do this, Pandoc's internal code now refers to a "dummy" counter called `none`. However, if you are using an **older LaTeX distribution** or a **custom template** (like *Eisvogel* or a custom `.tex` file) that hasn't been updated to define this counter, LaTeX gets confused because it doesn't know what `none` is.

---

### How to Fix It

You can resolve this using one of the following methods, ordered from easiest to most permanent:

#### Method 1: Add the counter to your Markdown (Easiest)

Add this small snippet to the **YAML metadata** at the very top of your Markdown file. This manually defines the missing counter before the PDF is generated.

```yaml
---
header-includes:
  - |
    \makeatletter
    \@ifundefined{c@none}{\newcounter{none}}{}
    \makeatother
---

```

#### Method 2: Use the `--variable` flag in your command

If you don't want to edit your Markdown file, you can "inject" the fix directly into your terminal command:

```bash
pandoc input.md -o output.pdf --variable include-before="\newcounter{none}"

```

#### Method 3: Update your Custom Template

If you are using a specific template (e.g., `--template mytemplate.tex`), you need to add the following line to the preamble of that `.tex` file (usually somewhere before `\begin{document}`):

```latex
\newcounter{none}

```

#### Method 4: Update Pandoc or your LaTeX Distro

* **Update Pandoc:** This issue was addressed in the default templates of **Pandoc 3.8.2.1+**. If you are on 3.8.1 or 3.8.2, upgrading might fix it if you aren't using a custom template.
* **Update TeX Live / MiKTeX:** The error sometimes stems from an outdated `longtable` package. Updating your LaTeX packages via your TeX manager (like TeX Live Utility or MiKTeX Console) can sometimes resolve the underlying conflict.

---

**Summary of the fix:** LaTeX is looking for a counter named `none` that Pandoc expects to exist. Defining it manually with `\newcounter{none}` satisfies the requirement and lets the PDF build.

**Would you like me to help you modify your specific Pandoc command or check if your Markdown table syntax is correct?**

**Comment by afrazahmad-hub** [comment]
This worked for me:

Tell Pandoc to stop trying to be fancy with the LaTeX `longtable` package. Use the `--columns flag` to force a standard layout.
I used following command:
`pandoc test.md -o report.pdf --pdf-engine=xelatex --columns=10`

**Reply by EngrPips** [reply]
Thanks for the heads-up, buddy.
